### Learning Portfolio 7

The project of this portolio is to solve a LP problem by using a new optimization model Pyomo.

In [7]:
pip install pyomo

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.6/10.6 MB 13.8 MB/s eta 0:00:0000:010:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.6/49.6 kB 1.9 MB/s eta 0:00:00
Note: you may need to restart the kernel to use updated packages.


In [79]:
from pyomo.environ import *
import pandas as pd

$$
\begin{array}{rc}
\text{Maximize:} & z = 3x_1 + 2x_2 + 4x_3  \\
\text{Subject to:} & x_1 + x_2 + 2x_3  \leqslant  4 \\
& 2x_1\qquad + 3x_3 \leqslant 5 \\
& 2x_1 + x_2 + 3x_3 \leqslant 7 \\
& x_1 \geq 0, x_2 \geq 0, x_3\geq 0
\end{array}
$$


### The creation of a space for solving LP problem

- By using the ConcreteModel() and assign as model, it gives us a white paper to start with the question.

In [86]:
model = ConcreteModel()

### The variables

- model.A: A is the name of variable adds to the model, here we add the non-negative real numnber to the model.
- Var: a pyomo function for making decision variables.
- Objective function: model.maximize by using Objective( exper = function, sense = maximize) to state the problem
- Constraints: 
    1. model.Constraint1 
    2. model.Constraint2  
    3. model.Constraint3 
    
    The name of constraints can be change manully.

In [89]:
#model creation
model = ConcreteModel()

#variable
model.A = Var(within = NonNegativeReals)
model.B = Var(within = NonNegativeReals)
model.C = Var(within = NonNegativeReals)

#set the objective function
model.maximize = Objective(expr = 3 * model.A + 2 * model.B + 4 * model.C, sense = maximize)

#set the constraints
model.Constraint1 = Constraint(expr = 1 * model.A + 1 * model.B + 2 * model.C <= 4)

model.Constraint2 = Constraint(expr = 2 * model.A + 3 * model.C <= 5)

model.Constraint3 = Constraint(expr = 2 * model.A + model.B + 3 * model.C <= 7)

#The solver
op = SolverFactory("glpk")
op.solve(model) 

#display the answer
model.display()

Model unknown

  Variables:
    A : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :   2.5 :  None : False : False : NonNegativeReals
    B : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :   1.5 :  None : False : False : NonNegativeReals
    C : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :   0.0 :  None : False : False : NonNegativeReals

  Objectives:
    maximize : Size=1, Index=None, Active=True
        Key  : Active : Value
        None :   True :  10.5

  Constraints:
    Constraint1 : Size=1
        Key  : Lower : Body : Upper
        None :  None :  4.0 :   4.0
    Constraint2 : Size=1
        Key  : Lower : Body : Upper
        None :  None :  5.0 :   5.0
    Constraint3 : Size=1
        Key  : Lower : Body : Upper
        None :  None :  6.5 :   7.0


As the answer shown, the optimal value is 10.5 and for $x_{1} = 2.5$, $x_{2} = 1.5$, $x_{3} = 0.0$

Learning Referece: 
- $\href{http://edge.rit.edu/content/P18751/public/Google%20drive%20backup/Pyomo%20-%20Optimization%20Modeling%20in%20Python%2C%20Second%20Edition.pdf}{Pyomo\ Optimization\ Modeling \ in\ Python}$

- $\href{https://medium.com/@chongjingting/4-ways-to-solve-linear-programming-in-python-b4af36b7894d}{4 \ Ways\  to\ Solve\ Linear\ Programming\ in\ Python}$
